### ISSUE TITLE
Trouble getting coordinates from a `SpatRaster`

### ISSUE THREAD

**Issue by Kodiologist** [rspatial/terra#338]
Don't get me wrong, `terra` is a great package, especially compared to `raster`. But I recently spent some time trying to figure out how to find the coordinates for each cell in a `terra` `SpatRaster`, and while I eventually found what I needed, I hit several stumbling blocks along the way:

- [This manual section](https://rspatial.org/terra/spatial/8-rastermanip.html#accessing-cell-values) mentions `getValues` and `getValuesBlock`, but no such functions are exported by `terra`. I presume this is a leftover from the `raster` version of this manual.
- [This manual section](https://rspatial.org/terra/pkg/8-cell_level_functions.html#accessing-cell-values) correctly names `values`, but `valuesBlock` doesn't seem to exist, either.
- The same section says that constructs like `r[1:3,1:3, drop=FALSE]` can "keep the matrix structure"; while this statement is arguably true, it's surprising that the result is another `SpatRaster` rather than a matrix.
- `?values` says "Use `as.matrix` for an alternative matrix representation where the number of rows and columns matches that of `x`, if `x` has a single layer." This only works if you set the parameter `wide`. The user probably wouldn't expect such a parameter to be required or allowed because it's not present in `base::as.matrix`. `?terra::as.matrix` shows the parameter exists, but doesn't explain what it does.
- The way to get coordinates for all cells turns out to be `xyFromCell(x, 1 : ncell(x))`. `xyFromCell(x)` isn't implemented and `xyFromCell(x, cells(x))` does the wrong thing because unary `cells` uses only non-NA cells. `cells` doesn't allow a parameter `na.rm` and the default behavior is to remove rather than keep NAs (unlike R base functions), which is unwise.

It might make sense to provide a function that returns a data frame with the coordinates in two columns (like `xyFromCell`) and the cell values in other columns (like `as.data.frame`).

**Comment by Rapsodia86** [comment]
How about just using terra::as.data.frame (page 26 from https://cran.r-project.org/web/packages/terra/terra.pdf)

> v <- vect(system.file("ex/lux.shp", package="terra"))[1:2,]
> z <- rast(v, res=.1)
> values(z) <- 1:ncell(z)
> 
> terra::as.data.frame(z, xy=T, cells=T, na.rm=TRUE)
   cell        x        y lyr.1
1     1 5.876232 50.13479     1
2     2 5.976232 50.13479     2
3     3 6.076232 50.13479     3
4     4 6.176232 50.13479     4
5     5 6.276232 50.13479     5
6     6 5.876232 50.03479     6
7     7 5.976232 50.03479     7
8     8 6.076232 50.03479     8
9     9 6.176232 50.03479     9
10   10 6.276232 50.03479    10
11   11 5.876232 49.93479    11
12   12 5.976232 49.93479    12
13   13 6.076232 49.93479    13
14   14 6.176232 49.93479    14
15   15 6.276232 49.93479    15
16   16 5.876232 49.83479    16
17   17 5.976232 49.83479    17
18   18 6.076232 49.83479    18
19   19 6.176232 49.83479    19
20   20 6.276232 49.83479    20

**Comment by Kodiologist** [comment]
Oh thanks, `as.data.frame` is another standard method that I didn't guess took additional arguments in the case of `SpatRaster`, particularly `xy` and `na.rm`. I notice that this `na.rm` defaults to true; I still think it should default to false.

**Comment by rhijmans** [comment]
Thank you very much for all these useful comments/corrections. I have improved the docs (the on-line manual not posted yet). 

To get coordinates, there is also the `crds` (coordinates) methods, and I have now added a `na.rm` argument to it.

I follow R behavior with `na.rm=FALSE` for summary statistics (even though that is a nuisance; it may help in detecting errors). 
I take your point, but It is not obvious to me that the same logic should apply to `as.data.frame` or `crds`.
